feat(lua): add minimal vim.net.download wrapper using curl

tampueroc opened the pull request:
This PR introduces a minimal `vim.net.download()` implementation as a simple wrapper around `curl`, following the approach suggested in [this comment](https://github.com/neovim/neovim/pull/29104#issuecomment-2165443624).

The goal is to start small and provide the foundational download functionality first and then add more advanced features (e.g. authentication, headers, proxy support) in future PRs.

This PR is related to issue #23232 

### Implementation

- Uses `vim.system()` with `curl` to perform the download.
- Automatically adds --retry <n>, --location (follow redirects) and --output <file> plus the URL.
- Accepts an `opts` table and an optional `on_exit` callback.
- If you pass an `on_exit` callback, it runs in the background and invokes `on_exit(err or nil)` when done.
- If no callback is given, it blocks until completion and returns `true` on success or `false, errmsg` on failure.

### Proposed API Usage

1. **Basic synchronous download**:
   ```lua
    local ok, err = vim.net.download("https://httpbingo.org/anything", "./anything.txt")
    if not ok then
      print("Download failed:", err)
    end
    ```
    
2. **Synchronous with custom options**:
    ```lua
   local ok, err = vim.net.download("https://httpbingo.org/anything", "./anything.txt", {
       verbose = true,
       retry = 5,
       }
   )
   if ok then
      print("Done")
    else
      print("Error:", err)
    end
    ```
   
3. **Asynchronous with callback**
    ```lua
    vim.net.download("https://httpbingo.org/anything", "./anything.txt", { retry = 2 }, function(err)
        if err then
            print("Download failed:", err)
        else
            print("Download successful!")
        end
    end)
    ```
    
### Next Steps

After initial feedback, the implementation can be extended incrementally, adding `vim.net.request()` or broader `curl` options support in follow-up PRs.

clason commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,30 @@
+
+  local args = {
+    'curl',
+    '--no-progress-meter',
```
maybe `--silent --fail --show-error` would be more appropriate here?

clason replied:
Do we want to add a small `--retry` to handle flakes? (CI, I'm looking at you...)

clason commented:
Needs docs and a news entry. It would help if the proposed API (for several representative use cases) was added to the PR description.

clason commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,30 @@
+local M = {}
+
+function M.download(url, opts)
```
I think the `location` should be a mandatory argument; I don't see a "put this _here_ (wherever 'here' is)" is a reasonable use case.

clason commented:
We should also add a health check for this (like we do for `rg`); see https://github.com/nvim-treesitter/nvim-treesitter/blob/e0d029bfdc1ea0da00356203b5afbe70ca9c74c1/lua/nvim-treesitter/health.lua#L87-L93

Here it's important to show the _full_ version output, including all build flags (since there are severely gimped `curl` versions out there that can lead to hard to diagnose problems).

clason commented:
You need to add the new module to https://github.com/neovim/neovim/blob/59c45b22d9616cda448b3c3dddd3e9274c830ca3/src/gen/gen_vimdoc.lua#L135

tampueroc commented:
> You need to add the new module to
> 
> https://github.com/neovim/neovim/blob/59c45b22d9616cda448b3c3dddd3e9274c830ca3/src/gen/gen_vimdoc.lua#L135

Thank you! I’ve been struggling to get the docs right (embarrassingly).

On my side I still have pending
- `verbose` flag (agree on this one for sure)
- `retry` flag (idem)

clason commented on runtime/doc/news.txt:
```diff
@@ -74,6 +74,7 @@ LSP
 
 LUA
 
+• Added |vim.net.download()|, a minimal cross-platform function for downloading files using curl. Supports an optional callback on completion.
```
```suggestion
• |vim.net.download()| allows downloading files.
```

tampueroc committed:
feat(lua): add minimal vim.net.download() using curl

Problem:
Neovim didn't have a simple built-in way to download files using system tools.
Earlier attempts to implement a full `vim.net` module tried to cover too much and stalled due to complexity.

Solution:
Add a minimal `vim.net.download()` function as a thin wrapper around `curl`.
Supports basic GET requests and saving to a user-specified location.
Includes an optional `on_exit` callback for reporting success or failure.

Also includes:
- Functional tests for both success and failure cases
- Documentation in lua.txt
- Healthcheck entry showing curl version and relevant environment variables
- New entry in news.txt under "LUA" section

tampueroc commented:
@clason Sorry to bother, but could you guide me a bit on the current CI issue? For some reason it's only failing on the `test / ubuntu asan clang unittest` unit tests...

zeertzjq commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,55 @@ local function check_external_tools()
+      -- Add subtitle only if any env var is present
+      local added_env_header = false
+      for _, var in ipairs({
+        'CURL_CA_BUNDLE',
+        'CURL_HOME',
+        'CURL_SSL_BACKEND',
+        'SSL_CERT_DIR',
+        'SSL_CERT_FILE',
+        'HTTPS_PROXY',
+        'HTTP_PROXY',
+        'ALL_PROXY',
+        'NO_PROXY',
```
From man page:
```
The environment variables can be specified in lower case or upper case. The lower case version has precedence.
```

lewis6991 commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,53 @@
+local M = {}
+
+--- Downloads a file from the given URL to a specified location.
+---
+--- @param url string The URL to download from.
+--- @param download_location string The file path where the content should be saved.
+--- @param opts? table Optional parameters:
+---   - on_exit (function|nil): Called when the download completes or fails. Receives error string or nil.
+---   - verbose (boolean|nil): Enables verbose curl output.
+---   - retry (integer|nil): Number of times to retry on transient failures (default: 3).
+---
+--- @return nil
+function M.download(url, download_location, opts)
```
Please make the callback the last positional argument.

```suggestion
--- Downloads a file from the given URL to a specified location.
---
--- @param url string The URL to download from.
--- @param output string The file path where the content should be saved.
--- @param opts? table Optional parameters:
---   - verbose (boolean|nil): Enables verbose curl output.
---   - retry (integer|nil): Number of times to retry on transient failures (default: 3).
--- @param on_exit fun(err?: string)
--- @return nil
function M.download(url, output, opts, on_exit)
```

We may also want this function to operate synchronously if no callback is given.

Also I think we may want a variation of this function where the output is passed back as a string. For this case `on_exit` should have type `fun(data?: string, err?: string)`

tampueroc replied:
I will add the synchronous behaviour, thank you.

> Also I think we may want a variation of this function where the output is passed back as a string. For this case on_exit should have type fun(data?: string, err?: string)

Yes, I believe we should implement this in a future PR as a separate function, something like `vim.net.request()`.

lewis6991 commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,53 @@
+    url,
+  })
+
+  vim.system(args, { text = true }, function(res)
+    if res.code ~= 0 or res.stderr ~= '' then
+      if opts.on_exit then
+        opts.on_exit(res.stderr ~= '' and res.stderr or 'Download failed')
+      end
+      return
+    end
+    if opts.on_exit then
+      opts.on_exit(nil)
+    end
+  end)
```
Perhaps add a sync form.

```suggestion
  local result

  vim.system(args, { text = true }, function(res)
    result = res
    if res.code ~= 0 or res.stderr ~= '' then
      if opts.on_exit then
        opts.on_exit(res.stderr ~= '' and res.stderr or 'Download failed')
      end
      return
    end
    if opts.on_exit then
      opts.on_exit(nil)
    end
  end)
  
  if on_exit then
    return
  end
  
  vim.wait(TIMEOUT, function()
    return result ~= nil
  end)
  
  if result.code ~= 0 then
    return nil, result.stdout or 'Download failed'
  end
  
  return result.stdout
```

miroshQa commented:
Shouldn't something like this at least support a progress report and cancellation mechanism? Otherwise, plugin authors will end up reinventing a more functional version of this function anyway

clason commented:
As a plugin author, I don't need it. Let's keep this to MVP.

zeertzjq commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,56 @@ local function check_external_tools()
+      -- Add subtitle only if any env var is present
+      local added_env_header = false
+      for _, var in ipairs({
+        'CURL_CA_BUNDLE',
+        'CURL_HOME',
+        'CURL_SSL_BACKEND',
+        'SSL_CERT_DIR',
+        'SSL_CERT_FILE',
+        'HTTPS_PROXY',
+        'HTTP_PROXY',
+        'ALL_PROXY',
+        'NO_PROXY',
```
You need to change these to lower-case, otherwise `var:upper` does nothing.

mike325 commented:
would it be possible to also introduce an auth flag via either to authenticate using either basic auth or via token ?
Either that or an option to provide extra arguments directly to `curl`

clason commented:
No, at least not in this PR. You can (and should) use environment variables for local configuration.

tampueroc commented:
Hello, in the last push I implemented as per the feedback:
- async/sync behaviour based on the `on_exit` argument along the corresponding functional tests
- health check for env vars upper or lowercase as per curl documentation

I see that is failing on `test/functional/terminal/tui_spec.lua` for mac os intel, any help with that?

brianhuster commented:
> for the (internal!) purpose of downloading spell files, WASM parsers

Sorry if this is off-topic, but does that mean Nvim has plan to include the plugin `nvim-treesitter` by default?

lewis6991 commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,67 @@
+  local args = { 'curl' }
+
+  if opts.verbose then
+    table.insert(args, '--verbose')
```
What effect does this have? Wouldn't the verbose output be in `res.stdout`?

justinmk commented:
As mentioned in https://github.com/neovim/neovim/pull/34140#issuecomment-2984316759 , I suggest starting with a `vim.net.fetch()` which, at first, might only do the basic "download" case, while keep the door open to expand it later. 

The parameters of the proposed request() and download() already look nearly the same. So that suggests that we have a basic interface that allows future expansion. And they don't need to be separate functions.

clason commented:
>  I suggest starting with a vim.net.fetch() which, at first, might only do the basic "download" case, while keep the door open to expand it later.

But isn't that exactly what this PR does (modulo `fetch` vs. `download`) naming? I don't understand why you closed this.

tampueroc commented:
Hi all, I'm closing this PR in favor of #34140, which consolidates download and fetch into a single vim.net.request() API. Thanks for the feedback here, please continue discussion over there!
